StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1135UsingDirectivesMustBeQualified.cs

```diff
@@ -0,0 +1,115 @@
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(ReadabilityResources.SA1135Description), ReadabilityResources.ResourceManager, typeof(ReadabilityResources));
+        private static readonly string HelpLink = "https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1135.md";
+
+        public static DiagnosticDescriptor DescriptorNamespace { get; } =
+            new DiagnosticDescriptor(DiagnosticId, Title, MessageFormatNamespace, AnalyzerCategory.ReadabilityRules, DiagnosticSeverity.Warning, AnalyzerConstants.EnabledByDefault, Description, HelpLink);
+
+        public static DiagnosticDescriptor DescriptorType { get; } =
+            new DiagnosticDescriptor(DiagnosticId, Title, MessageFormatType, AnalyzerCategory.ReadabilityRules, DiagnosticSeverity.Warning, AnalyzerConstants.EnabledByDefault, Description, HelpLink);
+
+        /// <inheritdoc/>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(DescriptorNamespace);
+
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(HandleCompilationStart);
+        }
+
+        private static void HandleCompilationStart(CompilationStartAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(HandleUsingDeclaration, SyntaxKind.UsingDirective);
+        }
+
+        private static void HandleUsingDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var usingDirective = (UsingDirectiveSyntax)context.Node;
+            CheckUsingDeclaration(context, usingDirective);
+        }
+
+        private static void CheckUsingDeclaration(SyntaxNodeAnalysisContext context, UsingDirectiveSyntax usingDirective)
+        {
+            if (!usingDirective.Parent.IsKind(SyntaxKind.NamespaceDeclaration))
+            {
+                // Usings outside of a namespace are always qualified.
+                return;
+            }
+
+            if (!usingDirective.StaticKeyword.IsKind(SyntaxKind.None))
+            {
+                // using static types is not considered.
+                return;
+            }
+
+            string usingString = usingDirective.Name.ToString();
+            if (usingString.IndexOf("::", StringComparison.Ordinal) >= 0)
+            {
+                // global qualified namespaces are OK.
+                return;
+            }
+
+            var symbol = context.SemanticModel.GetSymbolInfo(usingDirective.Name, context.CancellationToken).Symbol;
+            if (symbol == null)
+            {
+                // if there is no symbol, do not proceed.
+                return;
+            }
+
+            string symbolString = symbol.ToString();
```

Comment:
💭 It seems like this is prone to errors for different forms of generic types. In the tests, we covered `Dictionary<int, string>`. What happens if the user wrote `Collections.Dictionary<Int32, String>` instead? It should be easy to create unit tests which reveal the holes in this, and then file issues to resolve it in the future.

Reply:
See #2619